lib/puppet/type/cisco_itd_service.rb

```diff
@@ -0,0 +1,338 @@
+# limitations under the License.
+Puppet::Type.newtype(:cisco_itd_service) do
+  @doc = "Manages a Cisco ItdService.
+
+  Any resource dependency should be run before the interface resource.
+
+  cisco_itd_service {\"<name>\":
+    ..attributes..
+  }
+
+  <name> is the complete name of the service.
+
+  Example:
+    cisco_itd_service {\"my_service\":
+     ensure                        => present,
+     access_list                   => \"my_access1\",
+     device_group                  => \"my_group\",
+     exclude_access_list           => \"my_access2\",
+     fail_action                   => true,
+     ingress_interface             => [['vlan 2', '1.1.1.1'], ['ethernet 1/1', '2.2.2.2']],
+     load_bal_buckets              => 256,
+     load_bal_enable               => true,
+     load_bal_mask_pos             => 5,
+     load_bal_method_bundle_hash   => \"ip-l4port\",
+     load_bal_method_bundle_select => \"src\",
+     load_bal_method_end_port      => 100,
+     load_bal_method_proto         => \"tcp\",
+     load_bal_method_start_port    => 50,
+     nat_destination               => false,
+     peer_local                    => \"ser\",
+     peer_vdc                      => ['vdc1', 'ser'],
+     shutdown                      => true,
+     virtual_ip                    => ['ip 1.1.1.1 2.2.2.2', 'ip 2.2.2.2 255.0.0.0 udp 1000 device-group myGroup1'],
+    }"
+
+  ###################
+  # Resource Naming #
+  ###################
+
+  # Parse out the title to fill in the attributes in these
+  # patterns. These attributes can be overwritten later.
+  def self.title_patterns
+    identity = ->(x) { x }
+    patterns = []
+
+    # Below pattern matches both parts of the full composite name.
+    patterns << [
+      /^(\S+)$/,
+      [
+        [:itdser, identity]
+      ],
+    ]
+    patterns
+  end
+
+  newparam(:itdser, namevar: :true) do
+    desc "Name of the itd service. Valid value is a string of
+          non-whitespace characters. It is case-sensitive"
+    munge(&:strip)
+  end # param name
+
+  ##############
+  # Attributes #
+  ##############
+
+  ensurable
+
+  newproperty(:access_list) do
+    desc "ITD access-list name. Valid values are string, keyword
+         'default'. "
+
+    munge do |value|
+      value = :default if value == 'default'
+      value
+    end
+  end # property access_list
+
+  newproperty(:device_group) do
+    desc "ITD device-group name. Valid values are string, keyword
+         'default'. "
+
+    munge do |value|
+      value = :default if value == 'default'
+      value
+    end
+  end # property device_group
+
+  newproperty(:exclude_access_list) do
+    desc "ITD exclude-access-list name. Valid values are string, keyword
+         'default'. "
+
+    munge do |value|
+      value = :default if value == 'default'
+      value
+    end
+  end # property exclude_access_list
+
+  newproperty(:fail_action) do
+    desc 'ITD failaction'
```

Comment:
Should you provide a bit more info here for `fail_action`?  Not clear how this should be used.